pydra/utils/hash.py

```diff
@@ -51,19 +53,109 @@
+    _hashes: ty.Dict[CacheKey, Hash] = attrs.field(factory=dict)
+
+    @location.validator
+    def location_validator(self, _, location):
+        if not os.path.isdir(location):
+            raise ValueError(
+                f"Persistent cache location '{location}' is not a directory"
+            )
+
+    def get_hash(self, key: CacheKey, calculate_hash: ty.Callable) -> Hash:
+        """Check whether key is present in the persistent cache store and return it if so.
+        Otherwise use `calculate_hash` to generate the hash and save it in the persistent
+        store.
+
+        Parameters
+        ----------
+        key : CacheKey
+            locally unique key (e.g. to the host) used to lookup the corresponding hash
+            in the persistent store
+        calculate_hash : ty.Callable
+            function to calculate the hash if it isn't present in the persistent store
+
+        Returns
+        -------
+        Hash
+            _description_
+        """
+        try:
+            return self._hashes[key]
+        except KeyError:
+            pass
+        key_path = self.location / blake2b(str(key).encode()).hexdigest()
+        with SoftFileLock(key_path.with_suffix(".lock")):
+            if key_path.exists():
+                return Hash(key_path.read_bytes())
+            hsh = calculate_hash()
+            key_path.write_bytes(hsh)
+        return Hash(hsh)
+
+
+def persistent_cache_converter(
+    path: ty.Union[Path, str, PersistentCache, None]
+) -> PersistentCache:
+    if isinstance(path, PersistentCache):
+        return path
+    if path is None:
+        path = tempfile.mkdtemp()
+    return PersistentCache(Path(path))
+
+
+@attrs.define
+class Cache:
+    persistent: ty.Optional[PersistentCache] = attrs.field(
+        default=None,
+        converter=persistent_cache_converter,
+    )
+    _hashes: ty.Dict[int, Hash] = attrs.field(factory=dict)
+
+    def __getitem__(self, object_id: int) -> Hash:
+        return self._hashes[object_id]
+
+    def __setitem__(self, object_id: int, hsh: Hash):
+        self._hashes[object_id] = hsh
+
+    def __contains__(self, object_id):
+        return object_id in self._hashes
 
 
 class UnhashableError(ValueError):
     """Error for objects that cannot be hashed"""
 
 
-def hash_function(obj):
+def hash_function(obj, persistent_cache: ty.Optional[Path] = None):
     """Generate hash of object."""
-    return hash_object(obj).hex()
+    if persistent_cache is None:
+        # FIXME: Ideally the default location would be inside one of the cache_locations
+        # but can't think of a clean way to pass the cache_locations down to this part
+        # of the code, so just dumping in the home directory instead
+        persistent_cache = (Path("~") / ".pydra-hash-cache").expanduser()
+        try:
+            if not persistent_cache.exists():
+                persistent_cache.mkdir()
+        except Exception:
+            persistent_cache = None
```

Comment:
I believe we should be careful where to place stuff in the home directory.

We could follow the old "wild west" convention of placing files in a single `.pydra` folder at the root of `$HOME`, but this is generally considered sloppy, and sooner or later someone will complain and ask whether we can support established standards for their platform of choice.

Luckily, there is a [package](https://pypi.org/project/platformdirs/) implementing this already. I guess we would be concerned by this [section](https://platformdirs.readthedocs.io/en/latest/api.html#cache-directory).

Reply:
- perhaps we could just use a specific environmental variable if specified by the user?
- Since this is meant to be used within specific run, should we have option to clean the files that are e.g. a week old?

Reply:
That package looks good @ghisvail, although I not sure this is worth adding another dependency for, what are everyone's thoughts?

In either case, an env var is a good idea @djarecka. Do we have a naming convention for them?

> Since this is meant to be used within specific run, should we have option to clean the files that are e.g. a week old?

I don't see this needing to be constrained to a specific run, but probably no harm to run a clean-up of week/month-old files after a workflow completes

Reply:
> although I not sure this is worth adding another dependency for

Every piece of software which cache stuff to home end up writing (usually badly) some subset of what's covered in `platformdirs`. It's not worth adding if we are confident to implement it *right*. If not, the package is dependency free and [well tested](https://github.com/platformdirs/platformdirs/actions/workflows/check.yml), so the risks and additional bloat to Pydra would be minimal.

> we could just use a specific environmental variable if specified by the user

We could indeed (say `PYDRA_CACHE_DIR`), but the suitable fallback value is platform dependent. That's where the value of `platformdirs` comes in.

> I don't see this needing to be constrained to a specific run, but probably no harm to run a clean-up of week/month-old files after a workflow completes

I am trying to picture a use case for this: if I run workflow A and months later workflow B, should I expect Pydra to clean cached files up from workflow A (and any other which are old enough w.r.t. workflow B)?

Reply:
> > although I not sure this is worth adding another dependency for
> 
> Every piece of software which cache stuff to home end up writing (usually badly) some subset of what's covered in `platformdirs`. It's not worth adding if we are confident to implement it _right_. If not, the package is dependency free and [well tested](https://github.com/platformdirs/platformdirs/actions/workflows/check.yml), so the risks and additional bloat to Pydra would be minimal.

Sounds ok to me, I looked into pulling out just the bits we need from that package, and for Mac+Linux it was trivial but Windows is pretty complicated so decided against it.

> > we could just use a specific environmental variable if specified by the user
> 
> We could indeed (say `PYDRA_CACHE_DIR`), but the suitable fallback value is platform dependent. That's where the value of `platformdirs` comes in.

👍

> 
> > I don't see this needing to be constrained to a specific run, but probably no harm to run a clean-up of week/month-old files after a workflow completes
> 
> I am trying to picture a use case for this: if I run workflow A and months later workflow B, should I expect Pydra to clean cached files up from workflow A (and any other which are old enough w.r.t. workflow B)?

Yes, if they are not touched by Workflow B and they are over a month old they get cleaned up.

Reply:
It is still showing up as "1 change requested" from you, I think it might have been the method name change (which I implemented in the last round of changes). Is that right?

Reply:
@ghisvail I realised that *platformdirs* is already dependency of *black* and *virtualenv* so have used it to locate the hash cache as you suggested.